collections/_posts/2024-08-26-SAM2.md

```diff
@@ -0,0 +1,170 @@
+
+The image encoder is run only once over all the images to produce *unconditioned image tokens* representing each frame.
+
+&nbsp;
+
+## Memory attention
+
+The role of memory attention is to **condition frame features on the past feature frames and predictions** as well as on any new prompts. It is composed of several transformer blocks in which there is self-attention between the image embeddings of the current frame, and cross-attention to tokens of the objects in the memory bank, that is prompted or unprompted frames and objets pointers (see relevant section for further details).
+
+&nbsp;
+
+## Prompt encoder
+
+Prompt encoders are similar to SAM's : sparse prompts, *i.e.* points and bounding boxes, are represented by positional encodings summed with learnt embeddings for each prompt type and dense prompts, *i.e.* masks, are embedded using (downsampling) convolutions and summed with the conditioned frame embeddings.
+
+&nbsp;
+
+## Mask decoder
+
+The decoder architecture is again quite similar to that of SAM. It takes as input the conditioned frame embeddings (with potential mask encoded embeddings) on one hand, and sparse prompts tokens, concatenated with tokens used for the output, on the other hand. The decoder stacks transformer blocks with self-attention between prompt+ouput tokens and alternated "image to token" and "token to image" cross-attention.
+
+> N.B: X to Y attention means that X is query and Y is key and value in the attention layer.
+
+The model produces multiple from the resulting tokens : 
+* The tokens are upsampled with convolutions and summed with stages 1 and 2 frame features from Hiera image encoder. They are multiplied with the output tokens (after passing through 3 MLP layers) to produce **low-resolution masks** (at a resolution 4x lower than input image).
+*  The output tokens is also used to produce what are called **object pointers**, which are lightweight vectors to encode high-level semantic information of the objecct to segment.
+*  As in SAM, there could be ambiguity on the mask to predict, as an object can have several subparts that are also segmented. For this reason, the model produces multiple output masks (3 in general) and predicts the **IoU scores** between the ground truth and each mask.
+*  Finally, a specific constraint of video segmentation is that objects of interest can disappear from the image, momentarily or permanently. Hence, SAM 2 also predicts an **oclusion score** for each object.
+
+From these outputs, few post-processing steps, including resolving mask covering (based on the predicted IoU), mask refinement and a final 4x upsampling, allow to produce the final segmentation map at the original image size.
+
+&nbsp;
+
+## Memory bank
```

Comment:
If I understand the paper correctly, there are 2 FIFO queues: one of the _N_ most recent frames, and one of the _M_ most recent prompted frames. So I would try to rework the paragraph below to make this clear.

Reply:
see latest commit